src/main/java/com/amazonaws/encryptionsdk/keyrings/DecryptionMaterials.java

```diff
@@ -0,0 +1,189 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except
+ * in compliance with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package com.amazonaws.encryptionsdk.keyrings;
+
+import com.amazonaws.encryptionsdk.CryptoAlgorithm;
+
+import javax.crypto.SecretKey;
+import java.security.PublicKey;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.apache.commons.lang3.Validate.isTrue;
+import static org.apache.commons.lang3.Validate.notNull;
+
+/**
+ * Contains the cryptographic materials needed for a decryption operation with Keyrings.
+ */
+public final class DecryptionMaterials {
+    private final CryptoAlgorithm algorithm;
+    private SecretKey plaintextDataKey;
+    private final PublicKey verificationKey;
+    private final Map<String, String> encryptionContext;
+    private final KeyringTrace keyringTrace;
+
+    private DecryptionMaterials(Builder b) {
+        notNull(b.algorithm, "algorithm is required");
+        notNull(b.keyringTrace, "keyringTrace is required");
+        validatePlaintextDataKey(b.algorithm, b.plaintextDataKey);
+        validateVerificationKey(b.algorithm, b.verificationKey);
+
+        algorithm = b.algorithm;
+        plaintextDataKey = b.plaintextDataKey;
+        verificationKey = b.verificationKey;
+        encryptionContext = b.encryptionContext;
+        keyringTrace = b.keyringTrace;
+    }
+
+    /**
+     * The algorithm to use for this decryption operation.
+     */
+    public CryptoAlgorithm getAlgorithm() {
+        return algorithm;
+    }
+
+    public SecretKey getPlaintextDataKey() {
+        return plaintextDataKey;
+    }
+
+    /**
+     * Sets the plaintext data key. The plaintext data key must not already be populated.
+     *
+     * @param plaintextDataKey  The plaintext data key.
+     * @param keyringTraceEntry The keyring trace entry recording this action.
+     */
+    public void setPlaintextDataKey(SecretKey plaintextDataKey, KeyringTraceEntry keyringTraceEntry) {
+        if (this.plaintextDataKey != null) {
+            throw new IllegalStateException("plaintextDataKey was already populated");
+        }
+        notNull(plaintextDataKey, "plaintextDataKey is required");
+        notNull(keyringTraceEntry, "keyringTraceEntry is required");
+        validatePlaintextDataKey(algorithm, plaintextDataKey);
+        this.plaintextDataKey = plaintextDataKey;
+        keyringTrace.add(keyringTraceEntry);
+    }
+
+    public PublicKey getVerificationKey() {
+        return verificationKey;
```

Comment:
If `verificationKey` is `null` because the `algorithm` is not signed should this throw an error or just return `null`?

Reply:
Throwing an error would likely reduce incorrect usage of this class and EncryptionMaterials, so I've gone ahead and made the optional plaintextDataKey, verificationKey, and signingKey throw IllegalStateException if they are not populated. I've also added has* methods so that these properties can be checked if they are populated.